File: Asgmnt6/DataProcessor.go
package main

import (
	"fmt"
	"sort"
	"strings"
	"sync"
	"time"
)

/*
Multi-threaded Data Processing System
Copies 3 fixed lines from BufferA to BufferB using goroutines and channels
*/

// Configuration constants
const (
	WRITE_WORKERS   = 3
	READ_WORKERS    = 3
	PACKET_SIZE     = 8  // 8 characters per packet
	MAX_QUEUE_SIZE  = 10 // Channel buffer size
)

// Packet structure to hold data with metadata
type Packet struct {
	LineIndex      int
	SequenceNumber int
	Data           string
}

// String representation of packet for debugging
func (p Packet) String() string {
	return fmt.Sprintf("Packet{line=%d, seq=%d, data='%s'}", p.LineIndex, p.SequenceNumber, p.Data)
}

// Global data structures
var (
	bufferA = []string{
		"I'm not clumsy, it is the floor.",
		"Train'g is my only real defense.",
		"I'm on vegetable diet, I see it.",
	}
	bufferB    = make([]string, 3)
	packetMap  = make(map[int]map[int]string) // lineIndex -> (sequenceNumber -> data)
	mapMutex   sync.RWMutex                   // Protects packetMap
	
	// Channels for communication
	packetQueue    = make(chan Packet, MAX_QUEUE_SIZE)
	lineIndexChan  = make(chan int, len(bufferA))
	
	// Counters
	packetsProcessed int
	counterMutex     sync.Mutex
)

// Initialize system components
func initialize() {
	fmt.Println("=== Multi-threaded Data Processing System (using Go) ===")
	fmt.Printf("Copying 3 lines from BufferA to BufferB using %d write workers and %d read workers\n", 
		WRITE_WORKERS, READ_WORKERS)
	fmt.Println()
	
	// Initialize BufferB and packet map
	for i := 0; i < len(bufferB); i++ {
		bufferB[i] = ""
		packetMap[i] = make(map[int]string)
	}
	
	// Fill line index channel with available lines
	for i := 0; i < len(bufferA); i++ {
		lineIndexChan <- i
	}
	close(lineIndexChan)
	
	fmt.Println("BufferB initialized")
}

// Print BufferA contents
func printBufferA() {
	fmt.Println("\n=== BufferA (Original Data) ===")
	for i, line := range bufferA {
		fmt.Printf("Line %d (%d chars): \"%s\"\n", i, len(line), line)
	}
	fmt.Println()
}

// Write worker function - divides lines into packets and sends to channel
func writeWorker(workerID int, wg *sync.WaitGroup) {
	defer wg.Done()
	fmt.Printf("WriteWorker-%d started\n", workerID)
	
	for lineIndex := range lineIndexChan {
		line := bufferA[lineIndex]
		fmt.Printf("WriteWorker-%d processing line %d: \"%s\"\n", workerID, lineIndex, line)
		
		// Divide line into 8-character packets
		divideLineIntoPackets(line, lineIndex, workerID)
	}
	
	fmt.Printf("WriteWorker-%d finished\n", workerID)
}

// Divide a line into 8-character packets and send to channel
func divideLineIntoPackets(line string, lineIndex int, workerID int) {
	sequenceNumber := 0
	
	for start := 0; start < len(line); start += PACKET_SIZE {
		end := start + PACKET_SIZE
		if end > len(line) {
			end = len(line)
		}
		
		packetData := line[start:end]
		packet := Packet{
			LineIndex:      lineIndex,
			SequenceNumber: sequenceNumber,
			Data:           packetData,
		}
		
		// Send packet to channel (blocks if channel is full)
		packetQueue <- packet
		
		fmt.Printf("WriteWorker-%d added packet: \"%s\" (line %d, seq %d)\n", 
			workerID, packetData, lineIndex, sequenceNumber)
		
		sequenceNumber++
		
		// Small delay to simulate processing time
		time.Sleep(10 * time.Millisecond)
	}
}

// Read worker function - retrieves packets from channel and reconstructs BufferB
func readWorker(workerID int, wg *sync.WaitGroup) {
	defer wg.Done()
	fmt.Printf("ReadWorker-%d started\n", workerID)
	
	for packet := range packetQueue {
		// Process the packet
		processPacket(packet, workerID)
		
		// Update counter safely
		counterMutex.Lock()
		packetsProcessed++
		counterMutex.Unlock()
	}
	
	fmt.Printf("ReadWorker-%d finished\n", workerID)
}

// Process a packet and add it to the packet map
func processPacket(packet Packet, workerID int) {
	// Store packet in map by line and sequence number (thread-safe)
	mapMutex.Lock()
	packetMap[packet.LineIndex][packet.SequenceNumber] = packet.Data
	mapMutex.Unlock()
	
	fmt.Printf("ReadWorker-%d processed packet: \"%s\" (line %d, seq %d)\n", 
		workerID, packet.Data, packet.LineIndex, packet.SequenceNumber)
	
	// Small delay to simulate processing time
	time.Sleep(5 * time.Millisecond)
}

// Reconstruct BufferB from packet map in correct order
func reconstructBufferB() {
	mapMutex.RLock()
	defer mapMutex.RUnlock()
	
	for lineIndex := 0; lineIndex < len(bufferB); lineIndex++ {
		lineMap := packetMap[lineIndex]
		
		// Get sequence numbers and sort them
		var sequences []int
		for seq := range lineMap {
			sequences = append(sequences, seq)
		}
		sort.Ints(sequences)
		
		// Concatenate packets in correct order
		var builder strings.Builder
		for _, seq := range sequences {
			builder.WriteString(lineMap[seq])
		}
		
		bufferB[lineIndex] = builder.String()
	}
}

// Print BufferB contents
func printBufferB() {
	// First reconstruct BufferB from packet map
	reconstructBufferB()
	
	fmt.Println("=== BufferB (Copied Data) ===")
	for i, line := range bufferB {
		fmt.Printf("Line %d (%d chars): \"%s\"\n", i, len(line), line)
	}
	fmt.Println()
}

// Validate that BufferB matches BufferA
func validateBuffers() {
	fmt.Println("=== Buffer Validation ===")
	allMatch := true
	
	for i := 0; i < len(bufferA); i++ {
		matches := bufferA[i] == bufferB[i]
		if matches {
			fmt.Printf("Line %d: MATCHED \n", i)
		} else {
			fmt.Printf("Line %d: MISMATCHED \n", i)
			fmt.Printf("  Expected: \"%s\"\n", bufferA[i])
			fmt.Printf("  Got:      \"%s\"\n", bufferB[i])
			allMatch = false
		}
	}
	
	fmt.Println()
	if allMatch {
		fmt.Println("Result: All LINES MATCHED ")
	} else {
		fmt.Println("Result: VALIDATION FAILED ")
	}
	
	counterMutex.Lock()
	fmt.Printf("Total packets processed: %d\n", packetsProcessed)
	counterMutex.Unlock()
	fmt.Println()
}

// Run the complete data processing pipeline
func runDataProcessing() error {
	var writeWG, readWG sync.WaitGroup
	
	// Start write workers
	writeWG.Add(WRITE_WORKERS)
	for i := 1; i <= WRITE_WORKERS; i++ {
		go writeWorker(i, &writeWG)
	}
	
	// Start read workers
	readWG.Add(READ_WORKERS)
	for i := 1; i <= READ_WORKERS; i++ {
		go readWorker(i, &readWG)
	}
	
	// Wait for all write workers to complete
	writeWG.Wait()
	fmt.Println("All write workers completed")
	
	// Close packet queue to signal read workers to finish
	close(packetQueue)
	
	// Wait for all read workers to complete
	readWG.Wait()
	fmt.Println("All read workers completed")
	
	return nil
}

// Main function
func main() {
	startTime := time.Now()
	
	// Initialize system
	initialize()
	
	// Print original BufferA
	printBufferA()
	
	// Run processing pipeline
	err := runDataProcessing()
	if err != nil {
		fmt.Printf("Error during processing: %v\n", err)
		return
	}
	
	// Print results
	printBufferB()
	validateBuffers()
	
	// Show processing time
	endTime := time.Now()
	processingTime := endTime.Sub(startTime)
	fmt.Printf("Processing completed in %v\n", processingTime)
	fmt.Println("Data processing successful!")
}

File: Asgmnt6/DataProcessor.java
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.*;

/**
 * Multi-threaded Data Processing System
 * Copies 3 fixed lines from BufferA to BufferB using worker threads and shared queue
 */
public class DataProcessor {
    
    // Configuration
    private static final int WRITE_WORKERS = 3;
    private static final int READ_WORKERS = 3;
    private static final int PACKET_SIZE = 8; // 8 characters per packet
    private static final int MAX_QUEUE_SIZE = 10;
    
    // Shared data structures
    private static String[] bufferA = {
        "I'm not clumsy, it is the floor.",
        "Train'g is my only real defense.",
        "I'm on vegetable diet, I see it."
    };
    private static String[] bufferB = new String[3];
    private static Map<Integer, Map<Integer, String>> packetMap = new ConcurrentHashMap<>(); // lineIndex -> (sequenceNumber -> data)
    private static BlockingQueue<String> packetQueue = new LinkedBlockingQueue<>(MAX_QUEUE_SIZE);
    
    // Shared counters
    private static AtomicInteger nextLineIndex = new AtomicInteger(0);
    private static AtomicInteger packetsProcessed = new AtomicInteger(0);
    private static volatile boolean allPacketsAdded = false;
    
    /**
     * Main entry point
     */
    public static void main(String[] args) {
        System.out.println("=== Multi-threaded Data Processing System ===");
        System.out.println("Copying 3 lines from BufferA to BufferB using " + 
                          WRITE_WORKERS + " write workers and " + READ_WORKERS + " read workers");
        System.out.println();
        
        long startTime = System.currentTimeMillis();
        
        try {
            // Initialize BufferB
            initializeBufferB();
            
            // Print original BufferA
            printBufferA();
            
            // Start processing
            runDataProcessing();
            
            // Print results
            printBufferB();
            validateBuffers();
            
            long endTime = System.currentTimeMillis();
            System.out.println("Processing completed in " + (endTime - startTime) + " milliseconds");
            System.out.println(" Data processing successful!");
            
        } catch (Exception e) {
            System.err.println("Error during processing: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    /**
     * Initialize BufferB and packet map
     */
    private static void initializeBufferB() {
        for (int i = 0; i < bufferB.length; i++) {
            bufferB[i] = "";
            packetMap.put(i, new ConcurrentHashMap<>());
        }
        System.out.println("BufferB initialized");
    }
    
    /**
     * Run the complete data processing pipeline
     */
    private static void runDataProcessing() throws InterruptedException {
        // Create thread pools
        ExecutorService writeExecutor = Executors.newFixedThreadPool(WRITE_WORKERS);
        ExecutorService readExecutor = Executors.newFixedThreadPool(READ_WORKERS);
        
        try {
            // Start write workers
            for (int i = 0; i < WRITE_WORKERS; i++) {
                final int workerId = i + 1;
                writeExecutor.submit(() -> writeWorkerFunction(workerId));
            }
            
            // Start read workers
            for (int i = 0; i < READ_WORKERS; i++) {
                final int workerId = i + 1;
                readExecutor.submit(() -> readWorkerFunction(workerId));
            }
            
            // Wait for write workers to complete
            writeExecutor.shutdown();
            writeExecutor.awaitTermination(10, TimeUnit.SECONDS);
            allPacketsAdded = true;
            System.out.println("All write workers completed");
            
            // Wait for read workers to complete
            readExecutor.shutdown();
            readExecutor.awaitTermination(10, TimeUnit.SECONDS);
            System.out.println("All read workers completed");
            
        } finally {
            if (!writeExecutor.isShutdown()) {
                writeExecutor.shutdownNow();
            }
            if (!readExecutor.isShutdown()) {
                readExecutor.shutdownNow();
            }
        }
    }
    
    /**
     * Write worker function - divides lines into packets and adds to queue
     */
    private static void writeWorkerFunction(int workerId) {
        System.out.println("WriteWorker-" + workerId + " started");
        
        try {
            while (true) {
                // Get next line to process
                int lineIndex = nextLineIndex.getAndIncrement();
                
                if (lineIndex >= bufferA.length) {
                    break; // No more lines to process
                }
                
                String line = bufferA[lineIndex];
                System.out.println("WriteWorker-" + workerId + " processing line " + lineIndex + ": \"" + line + "\"");
                
                // Divide line into 8-character packets
                divideLineIntoPackets(line, lineIndex, workerId);
            }
            
        } catch (Exception e) {
            System.err.println("WriteWorker-" + workerId + " error: " + e.getMessage());
        }
        
        System.out.println("WriteWorker-" + workerId + " finished");
    }
    
    /**
     * Divide a line into 8-character packets and add to queue
     */
    private static void divideLineIntoPackets(String line, int lineIndex, int workerId) {
        try {
            for (int start = 0; start < line.length(); start += PACKET_SIZE) {
                int end = Math.min(start + PACKET_SIZE, line.length());
                String packet = line.substring(start, end);
                
                // Create packet with metadata: "lineIndex:sequenceNumber:data"
                int sequenceNumber = start / PACKET_SIZE;
                String packetWithMetadata = lineIndex + ":" + sequenceNumber + ":" + packet;
                
                // Add packet to queue (blocking if queue is full)
                packetQueue.put(packetWithMetadata);
                
                System.out.println("WriteWorker-" + workerId + " added packet: \"" + packet + 
                                 "\" (line " + lineIndex + ", seq " + sequenceNumber + ")");
                
                // Small delay to simulate processing time
                Thread.sleep(10);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.err.println("WriteWorker-" + workerId + " interrupted");
        }
    }
    
    /**
     * Read worker function - retrieves packets from queue and reconstructs BufferB
     */
    private static void readWorkerFunction(int workerId) {
        System.out.println("ReadWorker-" + workerId + " started");
        
        try {
            while (true) {
                // Get packet from queue with timeout
                String packetWithMetadata = packetQueue.poll(1, TimeUnit.SECONDS);
                
                if (packetWithMetadata == null) {
                    if (allPacketsAdded && packetQueue.isEmpty()) {
                        break; // No more packets to process
                    }
                    continue; // Keep waiting for more packets
                }
                
                // Process the packet
                processPacket(packetWithMetadata, workerId);
                packetsProcessed.incrementAndGet();
            }
            
        } catch (Exception e) {
            System.err.println("ReadWorker-" + workerId + " error: " + e.getMessage());
        }
        
        System.out.println("ReadWorker-" + workerId + " finished");
    }
    
    /**
     * Process a packet and add it to the packet map
     * Packets are stored by line and sequence number to ensure correct ordering
     */
    private static void processPacket(String packetWithMetadata, int workerId) {
        try {
            // Parse packet: "lineIndex:sequenceNumber:data"
            String[] parts = packetWithMetadata.split(":", 3);
            if (parts.length != 3) {
                System.err.println("Invalid packet format: " + packetWithMetadata);
                return;
            }
            
            int lineIndex = Integer.parseInt(parts[0]);
            int sequenceNumber = Integer.parseInt(parts[1]);
            String data = parts[2];
            
            // Store packet in map by line and sequence number
            Map<Integer, String> lineMap = packetMap.get(lineIndex);
            lineMap.put(sequenceNumber, data);
            
            System.out.println("ReadWorker-" + workerId + " processed packet: \"" + data + 
                             "\" (line " + lineIndex + ", seq " + sequenceNumber + ")");
            
            // Small delay to simulate processing time
            Thread.sleep(5);
            
        } catch (Exception e) {
            System.err.println("ReadWorker-" + workerId + " error processing packet: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    /**
     * Print BufferA contents
     */
    private static void printBufferA() {
        System.out.println("\n=== BufferA (Original Data) ===");
        for (int i = 0; i < bufferA.length; i++) {
            System.out.println("Line " + i + " (" + bufferA[i].length() + " chars): \"" + bufferA[i] + "\"");
        }
        System.out.println();
    }
    
    /**
     * Print BufferB contents
     */
    private static void printBufferB() {
        // Reconstruct BufferB from packet map in correct order
        for (int lineIndex = 0; lineIndex < bufferB.length; lineIndex++) {
            Map<Integer, String> lineMap = packetMap.get(lineIndex);
            StringBuilder lineBuilder = new StringBuilder();
            
            // Sort by sequence number and concatenate packets
            TreeMap<Integer, String> sortedPackets = new TreeMap<>(lineMap);
            for (String packet : sortedPackets.values()) {
                lineBuilder.append(packet);
            }
            
            bufferB[lineIndex] = lineBuilder.toString();
        }
        
        System.out.println("=== BufferB (Copied Data) ===");
        for (int i = 0; i < bufferB.length; i++) {
            System.out.println("Line " + i + " (" + bufferB[i].length() + " chars): \"" + bufferB[i] + "\"");
        }
        System.out.println();
    }
    
    /**
     * Validate that BufferB matches BufferA
     */
    private static void validateBuffers() {
        System.out.println("=== Buffers Validation ===");
        boolean allMatch = true;
        
        for (int i = 0; i < bufferA.length; i++) {
            boolean matches = bufferA[i].equals(bufferB[i]);
            System.out.println("Line " + i + ": " + (matches ? "MATCHED" : "MISMATCHED"));
            
            if (!matches) {
                System.out.println("  Expected: \"" + bufferA[i] + "\"");
                System.out.println("  Got:      \"" + bufferB[i] + "\"");
                allMatch = false;
            }
        }
        
        System.out.println();
        System.out.println("Overall Result: " + (allMatch ? "ALL LINES MATCHED" : "VALIDATION FAILED "));
        System.out.println("Total packets processed: " + packetsProcessed.get());
        System.out.println();
    }
}